File: Codes/1338-reduce-array-size-to-the-half.go
/*
 * @Author: LetMeFly
 * @Date: 2024-12-15 15:15:11
 * @LastEditors: LetMeFly.xyz
 * @LastEditTime: 2024-12-15 15:28:42
 */
package main
import "sort"

func minSetSize(arr []int) (ans int) {
    ma := map[int]int{}
    for _, t := range arr {
        ma[t]++
    }
    var times []int
    for _, t := range ma {
        times = append(times, t)
    }
    sort.Slice(times, func(i, j int) bool {
        return times[i] > times[j]
    })
    for cnt := 0; cnt < len(arr) / 2; ans++ {
        cnt += times[ans]
    }
    return
}

File: Codes/1338-reduce-array-size-to-the-half.py
'''
Author: LetMeFly
Date: 2024-12-15 15:06:44
LastEditors: LetMeFly.xyz
LastEditTime: 2024-12-15 15:09:07
'''
from typing import List
from collections import Counter

class Solution:
    def minSetSize(self, arr: List[int]) -> int:
        ma = Counter(arr)
        times = [t for i, t in ma.items()]
        times.sort()
        ans, cnt, i = 0, 0, len(times) - 1
        while cnt < len(arr) // 2:
            ans += 1
            cnt += times[i]
            i -= 1
        return ans